### PR TITLE
Separate mutable and immutable states

### LINKED ISSUE #6170: Separate mutable and immutable states
**Description**

While migrating the stream processors, we need to make sure that all state changes are reflected by events. To avoid unintended state changes on processing, we want to split the states into mutable and immutable parts. The immutable state will be used for reading the state on processing. The mutable state will be used for applying state changes of events.

### PR DESCRIPTION
**Pull request by saig0**
## Description

* extract mutable and immutable interfaces of the states
  * mutable interfaces are placed in the package `io.zeebe.engine.state.mutable`
  * immutable interfaces are placed in the package `io.zeebe.engine.state.immutable` 
  * mutable interfaces have the prefix `Mutable*State`
  * immutable interfaces have no prefix - following common conventions (e.g. like Scala/Kotlin collections)
  * the implementations have the prefix `Db*State` - to indicate that they are based on the `ZeebeDB`
* flatten the hierarchy of states
  * `ZeebeDb` give access to all states
  * avoid long call chains
* move methods to the responsible states
  * following the single responsibility principle
  * avoid feature envy 

**Hints for the Reviewer**

* no need to look over all changes - I just extracted the interfaces with the IDE
* no new JavaDoc for the interfaces - we can add the JavaDoc later iteratively when migrating the processors
* no additional improvements to keep the changes small
* the Sonar warnings exist before

## Related issues

closes #6170

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/state/mutable/MutableWorkflowState.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import io.zeebe.protocol.impl.record.value.deployment.DeploymentRecord;
+import org.agrona.DirectBuffer;
+
+public interface MutableWorkflowState extends WorkflowState {
+
+  void putDeployment(DeploymentRecord deploymentRecord);
+
+  void putLatestVersionDigest(DirectBuffer processId, DirectBuffer digest);
+
+  int getNextWorkflowVersion(String bpmnProcessId);
```
Comment (korthout) [inline engine/src/main/java/io/zeebe/engine/state/mutable/MutableWorkflowState.java:20]: Let's rename this to `nextWorkflowVersion` to make it more clear that this is not a getter without a side-effect

**Review by korthout — CHANGES_REQUESTED** [review]
These changes make me very happy. The naming convention feels really good. 

In the immutable variants I did have to think for some time about 2 methods that change internal state of the class, but not the actual state the class provides access to. Considering they don't actually change db state, its fine to keep them in the immutable interface.

I did find a method that should not be in an immutable interface (see comment) and I have a small naming suggestion.

Great work on this.

File: engine/src/main/java/io/zeebe/engine/state/immutable/EventScopeInstanceState.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  EventTrigger peekEventTrigger(long eventScopeKey);
+
+  /**
+   * Returns the next event trigger for the event scope or null if none exists. This will remove the
+   * polled event trigger from the state if it exists.
+   *
+   * @param eventScopeKey the key of the event scope
+   * @return the next event trigger or null if none exist
+   */
+  EventTrigger pollEventTrigger(long eventScopeKey);
```
[the remark this answers is not in the document]
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/state/immutable/EventScopeInstanceState.java:39]: Good catch :+1:

**Comment by saig0** [comment]
@korthout thanks for the fast review. I applied your hints. Please have a look :)

**Review by korthout — APPROVED** [review]
Thanks for the fast changes. LGTM 🥇

**Comment by saig0** [comment]
bors

**Pull request merged**